### PR TITLE
Development

### PR DESCRIPTION
**Pull request by sbebbers**
New release candidate

### CODE REVIEW

File: application/controller/DateController.php [hunk 1]
```diff
@@ -1,223 +1,230 @@
...
+        );
+        
+        $day    += $this->setDays((int)date('d'));
+        $month  += $this->setMonths('full', 'numeric', date('m'));
+        $year   += $this->setYears(1901, (int)date('Y'), 'asc', (int)date('Y'));
+        
+        $this->view->days   = $day;
+        $this->view->months = $month;
+        $this->view->years  = $year;    
+        end:
```
Comment (azvimba52) [inline application/controller/DateController.php:42]: do we need end: here?...oh goto...got it

File: application/core/FrameworkCore.php [hunk 2]
```diff
@@ -9,278 +10,283 @@
...
-						$this->setView($this->controller->{$_instance}->view);
-						$this->controller->{$_instance}->view = null;
-					}
-				}
-			}
+    public $segment, $host, $partial, $controller, $title, $description,
+    $serverPath, $root, $flash, $filePath, $uriPath, $http;
+    public $canonical   = '';
+    public $pageData    = [];
+    public $ignoredExts = [];
```
Comment (azvimba52) [inline application/core/FrameworkCore.php]: Probably better to declare all public attributes once i.e one line

File: application/core/FrameworkCore.php [hunk 3]
```diff
@@ -9,278 +10,283 @@
...
+        $this->allowedSegments  = $siteConfiguration['allowedSegments'];
+        $this->pageController   = $siteConfiguration['pageController'];
+        $this->errorReporting   = $siteConfiguration['errorReporting'] ?? [];
+        $this->allowedFileExts  = $siteConfiguration['allowedFileExts'] ?? [];
+        $this->ignoredExts      = $siteConfiguration['ignotedFileExts'] ?? ['js', 'css'];
+        
+        if(!empty($this->allowedSegments) && !empty($this->pageController)){
+            return true;
+        }
+        throw new FrameworkException("No pages or page controllers set in the pages.json file", "0x01");
```
Comment (azvimba52) [inline application/core/FrameworkCore.php:96]: I would do this the other way round ie if condition not met throw exception else return true

File: application/core/GlobalHelpers.php [hunk 4]
```diff
@@ -9,207 +8,208 @@
...
-		throw new FrameworkException("A site.json file is required in the configuration at the application level for the framework to run", "0x02");
-	}
-	if(empty($parameter)){
-		$parameter = 'baseURL';
-	}
-	return json_decode(file_get_contents(serverPath("/config/site.json")), true)[$parameter] ?? null;
+    if(!file_exists(serverPath("/config/site.json"))){
+        throw new FrameworkException("A site.json file is required in the configuration at the application level for the framework to run", "0x02");
+    }
+    if(empty($parameter)){
```
Comment (azvimba52) [inline application/core/GlobalHelpers.php]: $parameter defaults to 'cookieDomain' i.e never empty

Comment (sbebbers) [reply application/core/GlobalHelpers.php]: Good point well made. I'll look into it this evening.

File: application/core/GlobalHelpers.php [hunk 5]
```diff
@@ -9,207 +8,208 @@
...
- * @return	string
+ * @param   na
+ * @author  sbebbington
+ * @date    3 Feb 2017 09:04:29
+ * @version 0.1.4-RC3
+ * @return  string
  */
 function getServerIPAddress(){
-	return "{$_SERVER['SERVER_ADDR']}";
+    return "{$_SERVER['SERVER_ADDR']}";
```
Comment (azvimba52) [inline application/core/GlobalHelpers.php]: return $_SERVER['SERVER_ADDR'];

File: application/core/GlobalHelpers.php [hunk 6]
```diff
@@ -9,207 +8,208 @@
...
- * @return	string
+ * @param   na
+ * @author  sbebbington
+ * @date    2 Mar 2017 - 11:49:38
+ * @version 0.1.4-RC3
+ * @return  string
  */
 function getSelf(){
-	return $_SERVER['PHP_SELF'];
+    return "{$_SERVER['PHP_SELF']}";
```
Comment (azvimba52) [inline application/core/GlobalHelpers.php]: Again no need for quotes

File: application/core/FrameworkCore.php (hunk 3 above)
Comment (sbebbers) [reply application/core/FrameworkCore.php:96]: Maybe my mind works backwards.

File: application/controller/DateController.php (hunk 1 above)
Comment (sbebbers) [reply application/controller/DateController.php:42]: Haha - yes, don't you just love it.

On a serious note, now I've met the 'goto' challenge I was set, I'm going to not use it in future revision.

File: application/core/HtmlBuilder.php [hunk 7]
```diff
@@ -1,655 +1,656 @@
...
+     * the style attribute is added for use
+     * with jQuery and such
+     * 
+     * @param   int, string, string, string | array, string | array
+     * @author  sbebbington
+     * @date    30 Mar 2017 - 10:51:02
+     * @version 0.1.4-RC3
+     * @return  $this
+     */
+    public function h(int $size, string $text, string $id = '', $class = null, $style = null){
```
Comment (azvimba52) [inline application/core/HtmlBuilder.php:575]: Probably better not to use reserved words for function names i.e javascript, title, style etc

File: application/library/Library.php [hunk 8]
```diff
@@ -6,197 +6,196 @@
...
-		$shuffle	= str_shuffle("{$shuffle}");
-		
-		return substr($shuffle, 0, $numberToPad);		
-	}
+    private $key;
+    private $encryption;
+    
+    public function __construct(){
+        $this->key          = 'Skelet0n';
+        $this->encryption   = "AES-128-ECB";
```
Comment (azvimba52) [inline application/library/Library.php]: Probably better off in config file

**Comment by sbebbers** [comment]
All review points should be answered.

**Pull request merged**